MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java

```diff
@@ -441,37 +443,352 @@ private void maybeShowPopup(MouseEvent e) {
+                        JMenu m_e_menu = new JMenu(SkillType.ELITE_NM);
+
+                        // Aero Techs
+                        JMenu aero_techs = new JMenu("Aero Techs");
+                        JMenu a_ug_menu = new JMenu(SkillType.ULTRA_GREEN_NM);
+                        JMenu a_g_menu = new JMenu(SkillType.GREEN_NM);
+                        JMenu a_r_menu = new JMenu(SkillType.REGULAR_NM);
+                        JMenu a_v_menu = new JMenu(SkillType.VETERAN_NM);
+                        JMenu a_e_menu = new JMenu(SkillType.ELITE_NM);
+
+                        // Mechanics
+                        JMenu mechanics = new JMenu("Mechanics");
+                        JMenu v_ug_menu = new JMenu(SkillType.ULTRA_GREEN_NM);
+                        JMenu v_g_menu = new JMenu(SkillType.GREEN_NM);
+                        JMenu v_r_menu = new JMenu(SkillType.REGULAR_NM);
+                        JMenu v_v_menu = new JMenu(SkillType.VETERAN_NM);
+                        JMenu v_e_menu = new JMenu(SkillType.ELITE_NM);
+
+                        // BA Techs
+                        JMenu ba_techs = new JMenu("BA Techs");
+                        JMenu ba_ug_menu = new JMenu(SkillType.ULTRA_GREEN_NM);
+                        JMenu ba_g_menu = new JMenu(SkillType.GREEN_NM);
+                        JMenu ba_r_menu = new JMenu(SkillType.REGULAR_NM);
+                        JMenu ba_v_menu = new JMenu(SkillType.VETERAN_NM);
+                        JMenu ba_e_menu = new JMenu(SkillType.ELITE_NM);
+
                         for (Person tech : gui.getCampaign().getTechs()) {
-                        	if (tech.getMaintenanceTimeUsing() == 0 && !tech.isEngineer()) {
-                        		String skillLvl = "Unknown";
-                        		skillLvl = SkillType.getExperienceLevelName(tech.getExperienceLevel(false));
-                        		menuItem = new JMenuItem(tech.getFullTitle() + " (" + skillLvl + ", " + tech.getRoleDesc() + ")");
-                        		menuItem.setActionCommand("ADD_LANCE_TECH|FORCE|" + tech.getId() + "|" + forceIds);
-                        		menuItem.addActionListener(this);
-                        		menuItem.setEnabled(true);
-                        		menu.add(menuItem);
-                        		menu.setEnabled(true);
-                        	}
+                            if (tech.getMaintenanceTimeUsing() == 0 && !tech.isEngineer()) {
+                                String skillLvl = "Unknown";
+                                skillLvl = SkillType.getExperienceLevelName(tech.getExperienceLevel(tech.isTechPrimary() ? false : true));
+                                int role = tech.isTechPrimary() ? tech.getPrimaryRole() : tech.getSecondaryRole();
+
+                                // We're a MechTech!
+                                if (role == Person.T_MECH_TECH) {
+                                    if (skillLvl.equals(SkillType.ULTRA_GREEN_NM)) {
```

Comment:
The interiors of these if blocks are nearly identical except for which menu gets the new menu item. Can we condense the common functionality?

Reply:
That's premature optimization I don't want to do until post-stable. For now this works and I don't want to risk introduction of new bugs so close to the stable. Post-stable I plan on making sub-methods to handle the majority of this work.

Reply:
I'd argue that, since it's brand-new code, there's no greater risk of introducing bugs by making it more compact than there is of introducing copy + paste bugs.

Reply:
Except when you PR you're not supposed to just read, you're also supposed to compile and test. So, unless it tests incorrectly and you find it's due to a copy/paste bug... I'm inclined to leave it as is until I move it to utility methods post-stable. Right now it works (at least for me) perfectly and I don't want to optimize it when I know I'll be making many further changes post-stable. Hence premature optimization.

Reply:
So, the big reason not to optimize this before stable and the changes I plan on making post-stable, is that there will be more divergence then. I just don't know how much. I don't want to combine them, just to split them again in a couple months.

Reply:
I'll take your word for it.